sszgen/main.go

```diff
@@ -441,7 +474,53 @@ func (e *env) print(order []string, experimental bool) (string, bool, error) {
 		data["imports"] = importsStr
 	}
 
-	return execTmpl(tmpl, data), true, nil
+	code := execTmpl(tmpl, data)
+
+	result, err := removeUnusedImports(code, data["imports"].([]string))
+	if err != nil {
+		return "", false, err
+	}
+
+	return result, true, nil
+}
+
+func removeUnusedImports(code string, imports []string) (string, error) {
+	importUsed := make(map[string]bool)
+	for _, i := range imports {
+		// Import is of the form 'alias path'. We need only the alias.
+		importUsed[strings.Split(i, " ")[0]] = false
+	}
+	fset := token.NewFileSet()
+	f, err := parser.ParseFile(fset, "ssz.go", code, 0)
+	if err != nil {
+		return "", fmt.Errorf("failed to parse template: %w", err)
+	}
+	ast.Inspect(f, func(node ast.Node) bool {
+		switch n := node.(type) {
+		case *ast.SelectorExpr:
+			s := code[n.X.Pos()-1 : n.X.End()-1]
+			_, ok := importUsed[s]
+			if ok {
+				importUsed[s] = true
+			}
+		}
+		return true
+	})
+	ast.Inspect(f, func(node ast.Node) bool {
+		switch n := node.(type) {
+		case *ast.ImportSpec:
+			name := n.Name.String()
+			for i, used := range importUsed {
+				if name == i && !used {
+					// Add 1 to n.End() to remove trailing newline.
+					code = code[:n.Pos()-2] + code[n.End()+1:]
```

Comment:
Question: does this code edit the file in place or does it modify a copy? If it edits the file in place, then we should start at the end of the file and go backwards, otherwise removing more than 1 import won't work.

Reply:
What about importing the code that `goimports` calls? 

https://cs.opensource.google/go/x/tools/+/master:imports/forward.go;l=44

Note that it only tries to look for a file if the second argument is nil; it can take the source as a byte string and return the resulting source bytes after fixing imports. I think this also calls the same code as go fmt under the hood, so we could potentially even skip remove the call to `go/format` if the result is the same (w/ the added benefit of fixing imports).